File: frontend/src/lib/attachmentPolicy.ts
import type { AgentUIConfig } from '@/types/agent'

export type AttachmentPolicy = NonNullable<AgentUIConfig['attachments']>

const DEFAULT_MAX_MB = 10

/** First chunk for magic-byte sniffing (aligned with backend ``MAGIC_SNIFF_BYTES``). */
export const MAGIC_SNIFF_BYTES = 512

const OOXML_DOCX =
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
const OOXML_XLSX =
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
const OOXML_PPTX =
  'application/vnd.openxmlformats-officedocument.presentationml.presentation'

const ZIP_FAMILY = new Set<string>([
  'application/zip',
  OOXML_DOCX,
  OOXML_XLSX,
  OOXML_PPTX,
])

function bytesEq(head: Uint8Array, start: number, seq: readonly number[]): boolean {
  if (head.length < start + seq.length) return false
  for (let i = 0; i < seq.length; i++) {
    if (head[start + i] !== seq[i]) return false
  }
  return true
}

/** Best-effort MIME from leading bytes; ``null`` if unknown or too short. */
export function sniffMimeMagic(head: Uint8Array): string | null {
  if (head.length < 4) return null
  if (bytesEq(head, 0, [0x25, 0x50, 0x44, 0x46])) return 'application/pdf'
  if (bytesEq(head, 0, [0xff, 0xd8, 0xff])) return 'image/jpeg'
  if (bytesEq(head, 0, [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))
    return 'image/png'
  if (
    bytesEq(head, 0, [0x47, 0x49, 0x46, 0x38, 0x37, 0x61]) ||
    bytesEq(head, 0, [0x47, 0x49, 0x46, 0x38, 0x39, 0x61])
  ) {
    return 'image/gif'
  }
  if (
    head.length >= 12 &&
    bytesEq(head, 0, [0x52, 0x49, 0x46, 0x46]) &&
    bytesEq(head, 8, [0x57, 0x45, 0x42, 0x50])
  ) {
    return 'image/webp'
  }
  if (bytesEq(head, 0, [0x50, 0x4b, 0x03, 0x04])) return 'application/zip'
  if (head[0] === 0x4d && head[1] === 0x5a) return 'application/x-msdownload'
  return null
}

function mimeCompatible(declared: string, sniffed: string): boolean {
  const d = declared.toLowerCase()
  const s = sniffed.toLowerCase()
  if (d === s) return true
  if (ZIP_FAMILY.has(d) && ZIP_FAMILY.has(s)) return true
  if (
    (d === 'image/jpg' || d === 'image/jpeg') &&
    (s === 'image/jpg' || s === 'image/jpeg')
  ) {
    return true
  }
  return false
}

function executableDisguise(filename: string, sniffed: string | null): boolean {
  if (sniffed !== 'application/x-msdownload') return false
  const low = filename.toLowerCase()
  const safe = ['.exe', '.msi', '.dll', '.bat', '.cmd', '.scr']
  return !safe.some((s) => low.endsWith(s))
}

function extensionMagicConsistent(filename: string, sniffed: string | null): boolean {
  if (!sniffed) return true
  const low = filename.toLowerCase()
  if (low.endsWith('.pdf')) return sniffed === 'application/pdf'
  if (low.endsWith('.png')) return sniffed === 'image/png'
  if (low.endsWith('.jpg') || low.endsWith('.jpeg')) return sniffed === 'image/jpeg'
  if (low.endsWith('.gif')) return sniffed === 'image/gif'
  if (low.endsWith('.webp')) return sniffed === 'image/webp'
  if (low.endsWith('.docx')) return ZIP_FAMILY.has(sniffed)
  if (low.endsWith('.xlsx')) return ZIP_FAMILY.has(sniffed)
  if (low.endsWith('.pptx')) return ZIP_FAMILY.has(sniffed)
  return true
}

function matchesPattern(filename: string, mime: string, pattern: string): boolean {
  const p = pattern.trim()
  if (!p) return false
  const low = p.toLowerCase()
  if (low.startsWith('.')) {
    return filename.toLowerCase().endsWith(low)
  }
  if (low.includes('*') || low.includes('?')) {
    const main = mime.toLowerCase().split(';')[0]?.trim() ?? ''
    const escaped = low
      .replace(/[.+^${}()|[\]\\]/g, '\\$&')
      .replace(/\*/g, '.*')
      .replace(/\?/g, '.')
    const rx = new RegExp(`^${escaped}$`)
    return rx.test(main)
  }
  return mime.toLowerCase() === low
}

/** Whether the attachment button should be shown (default: enabled). */
export function isAttachmentUploadAllowed(policy: AttachmentPolicy | undefined): boolean {
  return policy?.enabled !== false
}

/** HTML ``accept`` attribute string when patterns are declared. */
export function attachmentInputAccept(policy: AttachmentPolicy | undefined): string | undefined {
  if (!policy?.accept?.length) return undefined
  const parts = policy.accept.map((x) => String(x).trim()).filter(Boolean)
  return parts.length ? parts.join(',') : undefined
}

/**
 * Client-side validation (size, accept, MIME magic) aligned with backend
 * ``attachment_policy.validate_upload_for_ui_config``.
 */
export async function validateLocalAttachment(
  file: File,
  policy: AttachmentPolicy | undefined,
): Promise<{ ok: true } | { ok: false; message: string }> {
  if (policy?.enabled === false) {
    return { ok: false, message: '该 Agent 未开启附件上传' }
  }
  const maxMb =
    typeof policy?.max_size_mb === 'number' && policy.max_size_mb > 0
      ? policy.max_size_mb
      : DEFAULT_MAX_MB
  const maxBytes = maxMb * 1024 * 1024
  if (file.size > maxBytes) {
    return { ok: false, message: `文件需不超过 ${maxMb}MB` }
  }

  let sniffed: string | null = null
  if (file.size > 0) {
    const slice = file.slice(0, Math.min(file.size, MAGIC_SNIFF_BYTES))
    const buf = new Uint8Array(await slice.arrayBuffer())
    sniffed = sniffMimeMagic(buf)
  }

  const declaredRaw = (file.type || '').trim().toLowerCase()
  const declared = declaredRaw || 'application/octet-stream'
  let effectiveMime = declared

  if (sniffed) {
    if (executableDisguise(file.name, sniffed)) {
      return { ok: false, message: '文件内容与类型不一致' }
    }
    if (!extensionMagicConsistent(file.name, sniffed)) {
      return { ok: false, message: '文件内容与类型不一致' }
    }
    if (
      declared === 'application/octet-stream' ||
      declared === 'binary/octet-stream' ||
      declared === ''
    ) {
      effectiveMime = sniffed
    } else if (!mimeCompatible(declared, sniffed)) {
      return { ok: false, message: '文件内容与类型不一致' }
    }
  }

  const accept = policy?.accept
  if (Array.isArray(accept) && accept.length > 0) {
    const patterns = accept.map((x) => String(x).trim()).filter(Boolean)
    if (patterns.length > 0) {
      const hit = patterns.some((pat) =>
        matchesPattern(file.name, effectiveMime, pat),
      )
      if (!hit) {
        return { ok: false, message: '不允许的文件类型' }
      }
    }
  }
  return { ok: true }
}

File: backend/src/agent_factory/core/attachment_policy.py
"""Upload constraints from Agent ``ui_config.attachments`` (docs/39)."""

from __future__ import annotations

import fnmatch
from typing import Any

DEFAULT_MAX_BYTES = 10 * 1024 * 1024

# First chunk used for magic-byte sniffing (plan §12 hardening).
MAGIC_SNIFF_BYTES = 512

_OOXML_DOCX = (
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
)
_OOXML_XLSX = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
_OOXML_PPTX = (
    "application/vnd.openxmlformats-officedocument.presentationml.presentation"
)

_ZIP_FAMILY = frozenset(
    {
        "application/zip",
        _OOXML_DOCX,
        _OOXML_XLSX,
        _OOXML_PPTX,
    }
)


def _attachments_section(ui_config: dict[str, Any] | None) -> dict[str, Any]:
    if not ui_config:
        return {}
    raw = ui_config.get("attachments")
    return raw if isinstance(raw, dict) else {}


def _max_bytes(section: dict[str, Any]) -> int:
    mb = section.get("max_size_mb")
    if isinstance(mb, (int, float)) and mb > 0:
        return int(mb * 1024 * 1024)
    return DEFAULT_MAX_BYTES


def sniff_mime_magic(head: bytes) -> str | None:
    """Best-effort MIME from leading bytes; ``None`` if unknown or too short."""
    if len(head) < 4:
        return None
    if head.startswith(b"%PDF"):
        return "application/pdf"
    if head.startswith(b"\xff\xd8\xff"):
        return "image/jpeg"
    if head.startswith(b"\x89PNG\r\n\x1a\n"):
        return "image/png"
    if head.startswith((b"GIF87a", b"GIF89a")):
        return "image/gif"
    if (
        len(head) >= 12
        and head.startswith(b"RIFF")
        and head[8:12] == b"WEBP"
    ):
        return "image/webp"
    if head.startswith(b"PK\x03\x04"):
        return "application/zip"
    if head.startswith(b"MZ"):
        return "application/x-msdownload"
    return None


def _mime_compatible(declared: str, sniffed: str) -> bool:
    d, s = declared.lower(), sniffed.lower()
    if d == s:
        return True
    if d in _ZIP_FAMILY and s in _ZIP_FAMILY:
        return True
    if d in ("image/jpg", "image/jpeg") and s in ("image/jpg", "image/jpeg"):
        return True
    return False


def _executable_disguise(filename: str, sniffed: str | None) -> bool:
    """True when PE executable magic but filename hides as document/image."""
    if sniffed != "application/x-msdownload":
        return False
    low = filename.lower()
    safe_suffix = (".exe", ".msi", ".dll", ".bat", ".cmd", ".scr")
    return not any(low.endswith(x) for x in safe_suffix)


def _extension_magic_consistent(filename: str, sniffed: str | None) -> bool:
    """Reject obvious extension lies when magic is confident."""
    if not sniffed:
        return True
    low = filename.lower()
    if low.endswith(".pdf"):
        return sniffed == "application/pdf"
    if low.endswith(".png"):
        return sniffed == "image/png"
    if low.endswith((".jpg", ".jpeg")):
        return sniffed == "image/jpeg"
    if low.endswith(".gif"):
        return sniffed == "image/gif"
    if low.endswith(".webp"):
        return sniffed == "image/webp"
    if low.endswith(".docx"):
        return sniffed in _ZIP_FAMILY
    if low.endswith(".xlsx"):
        return sniffed in _ZIP_FAMILY
    if low.endswith(".pptx"):
        return sniffed in _ZIP_FAMILY
    return True


def _matches_pattern(filename: str, mime_type: str, pattern: str) -> bool:
    p = pattern.strip()
    if not p:
        return False
    low = p.lower()
    if low.startswith("."):
        return filename.lower().endswith(low)
    if "*" in p or "?" in p:
        return fnmatch.fnmatch(mime_type.lower(), low)
    return mime_type.lower() == low


def validate_upload_for_ui_config(
    *,
    filename: str,
    mime_type: str,
    size_bytes: int,
    ui_config: dict[str, Any] | None,
    content_head: bytes | None = None,
) -> tuple[bool, str | None]:
    """Return (ok, error_code). ``error_code`` maps to API ``AgentFactoryException``.

    When ``content_head`` is set (typically first :data:`MAGIC_SNIFF_BYTES` bytes),
    declared MIME is cross-checked with magic bytes to reduce spoofing.
    """
    section = _attachments_section(ui_config)
    if section.get("enabled") is False:
        return False, "ATTACHMENTS_DISABLED"

    max_b = _max_bytes(section)
    if size_bytes > max_b:
        return False, "FILE_TOO_LARGE"

    declared_raw = (mime_type or "").strip().lower()
    declared = declared_raw or "application/octet-stream"
    sniffed: str | None = None
    if content_head:
        sniffed = sniff_mime_magic(content_head[:MAGIC_SNIFF_BYTES])

    effective_mime = declared
    if sniffed:
        if _executable_disguise(filename, sniffed):
            return False, "MIME_MAGIC_MISMATCH"
        if not _extension_magic_consistent(filename, sniffed):
            return False, "MIME_MAGIC_MISMATCH"
        if declared in ("application/octet-stream", "binary/octet-stream", ""):
            effective_mime = sniffed
        elif not _mime_compatible(declared, sniffed):
            return False, "MIME_MAGIC_MISMATCH"

    accept = section.get("accept")
    if isinstance(accept, list) and accept:
        patterns = [str(x).strip() for x in accept if str(x).strip()]
        if patterns:
            if not any(
                _matches_pattern(filename, effective_mime, pat) for pat in patterns
            ):
                return False, "FILE_TYPE_NOT_ALLOWED"

    return True, None
